eqcorrscan/tests/correlate_test.py

```diff
@@ -2,112 +2,395 @@
+def array_stream(array_template):
+    """
+    Return a stream genearted with randomn for testing normxcorr functions
     """
-    Run core tests
+    stream = random.randn(10000) * 5
+    # insert a template into the stream so cc == 1 at some place
+    ar = array_template[0]
+    stream[starting_index: starting_index + len(ar)] = ar
+    return stream
+
+
+@pytest.fixture(scope='module')
+def pads(array_template):
     """
-    unittest.main()
+    return an array of zeros for padding, matching templates len.
+    """
+    return np.zeros(array_template.shape[0], dtype=int)
+
+
+@pytest.fixture(scope='module')
+def array_ccs(array_template, array_stream, pads):
+    """  use each function stored in the normcorr cache to correlate the
+     templates and arrays, return a dict with keys as func names and values
+     as the cc calculated by said function"""
+    out = {}
+
+    for name in list(corr.XCOR_FUNCS.keys()):
+        func = corr.get_array_xcorr(name)
+        cc, _ = time_func(func, name, array_template, array_stream, pads)
+        out[name] = cc
+    return out
+
+
+# stream fixtures
+
+
+@pytest.fixture(scope='module')
+def multichannel_templates():
+    """ create multichannel templates """
+    return generate_multichannel_templates()
+
+
+@pytest.fixture(scope='module')
+def multichannel_stream():
+    """ create a multichannel stream for tests """
+    return generate_multichannel_stream()
+
+
+# a dict of all registered stream functions (this is a bit long)
+stream_funcs = {fname + '_' + mname: corr.get_stream_xcorr(fname, mname)
+                for fname in corr.XCOR_FUNCS.keys()
+                for mname in corr.XCORR_STREAM_METHODS
+                if fname != 'default'}
+
+
+@pytest.fixture(scope='module')
+def stream_cc_output_dict(multichannel_templates, multichannel_stream):
+    """ return a dict of outputs from all stream_xcorr functions """
+    # corr._get_array_dicts(multichannel_templates, multichannel_stream)
+    out = {}
+    for name, func in stream_funcs.items():
+        cc_out = time_func(func, name, multichannel_templates,
+                           multichannel_stream)
+        out[name] = cc_out
+    return out
+
+
+@pytest.fixture(scope='module')
+def stream_cc_dict(stream_cc_output_dict):
+    """ return just the cc arrays fro mthe stream_cc functions """
+    return {name: result[0] for name, result in stream_cc_output_dict.items()}
+
+
+# ----------------------------------- tests
+
+
+class TestArrayCorrelateFunctions:
+    """ these tests ensure the various implementations of normxcorr return
+    approximately the same answers """
+
+    # tests
+    def test_single_channel_similar(self, array_ccs):
+        """ ensure each of the correlation methods return similar answers
+        given the same input data """
+        cc_list = list(array_ccs.values())
+        for cc1, cc2 in itertools.combinations(cc_list, 2):
+            assert np.allclose(cc1, cc2, atol=0.05)
```

Comment:
Can we push this atol lower? With properly normalised time-domain cross-correlations they should all be *more* similar now?

Reply:
When I lower the atol the tests fail. Maybe this branch doesn't have the commit from https://github.com/eqcorrscan/EQcorrscan/pull/121? I was able to find the commits in the git log though... I am a bit confused on this one.

Reply:
How low did you go? Is it obvious which correlation doesn't match the others, or do they all fail?

Reply:
It was time-domain that was failing. Just had to delete the build directory and re-compile (I should have learned my lesson from https://github.com/eqcorrscan/EQcorrscan/issues/112)

Reply:
test now pass with atol = .00001, do you want it lower or is that sufficient?

Reply:
That looks like a good level - I have also made an issue to remind me to change how build works (#141), although I'm not sure how to do it, but it would make avoiding these issues much simpler, and I think you suggested it in #112 and I forgot to actually do it?